Reaction Ensemble refactor and unit testing

jngrad opened the pull request:
Fixes #4157

Description of changes:
- make `SingleReaction` class self-contained
- make RE methods and arguments `const` whenever possible
- write unit tests for the acceptance probability of all RE classes
- write integration tests to increase code coverage of RE to 90%

jngrad committed:
core: Move SingleReaction constructor

jngrad committed:
core: ReactionEnsemble: make everything const

jngrad committed:
tests: Unit test ReactionEnsemble classes

jngrad committed:
tests: Check factorial expression

jngrad committed:
core: Make acceptance probability methods protected

This way the methods can be unit tested.

jngrad committed:
tests: Check acceptance probability methods

jngrad committed:
core: Remove prototype without implementation

jngrad committed:
core: Add missing includes

jngrad committed:
core: Fix -Wconversion

KaiSzuttor commented on src/core/reaction_ensemble.hpp:
```diff
@@ -38,21 +40,41 @@
+                 std::vector<int> const &reactant_coefficients,
+                 std::vector<int> const &product_types,
+                 std::vector<int> const &product_coefficients) {
+    std::copy(reactant_types.begin(), reactant_types.end(),
```
why not just ```this->reactant_types = reactant_types```?

jngrad replied:
yes, these `std::copy` are unnecessary

KaiSzuttor commented on src/core/unit_tests/reaction_ensemble_classes_test.cpp:
```diff
@@ -0,0 +1,389 @@
+#include <memory>
+#include <stdexcept>
+#include <vector>
+
```
all the tests need documentation

jngrad replied:
I'm working on it

jngrad replied:
The unit test comments now show the equation of the chemical reactions and explain the different Monte Carlo schemes being tested. The particle types are also now stored in variables instead of hardcoded.

jngrad committed:
tests: Check Wang-Landau CV and histogram

jngrad committed:
core: Improve documentation of RE classes

jngrad committed:
test: Improve clarity of RE unit tests

jngrad committed:
core: Simplify RE code

jngrad committed:
tests: Rename Reaction Ensemble statistical tests

Also do some cosmetic changes.

jngrad committed:
tests: Split Wang-Landau test

Move the Wang-Landau output files checks to an interface test. The
statistical test runs the Monte Carlo algorithm until convergence.

jngrad committed:
tests: Create a faster constant pH test

Sample at a pH farther away from the pKa, shorten warmup phase,
run with 1 MPI rank to remove large overhead from communication.

jngrad committed:
tests: Remove UB in unit test

KaiSzuttor commented on src/core/unit_tests/reaction_ensemble_classes_test.cpp:
```diff
@@ -0,0 +1,499 @@
+    // check Wang-Landau histogram
+    {
+      r_algo.write_wang_landau_results_to_file(filename_output);
+      std::ifstream outfile;
```
please avoid writing to disk in the tests

jngrad replied:
This check and the corresponding RE code was changed to use a file-like object (`stringstream`) instead of an output file. The collective variable however still uses an input file that is created at the start of the test; the collective variable and Wang-Landau classes are so intertwined, that adapting them to take a file-like object would require a lot of refactoring.

jngrad replied:
After more refactoring, the unit test no longer depends on files.

jngrad committed:
core: Refactor Wang-Landau write function

Separate the formatting code from the file opening code.
Helps with unit testing.

jngrad committed:
tests: Simplify code

KaiSzuttor commented on src/core/reaction_ensemble.cpp:
```diff
@@ -69,26 +78,22 @@ void EnergyCollectiveVariable::load_CV_boundaries(
+    auto it = values.end();
+    wl_system.max_boundaries_energies.push_back(*(--it));
+    wl_system.min_boundaries_energies.push_back(*(--it));
+  }
```
if you'd extract this part as a function returning the values, we could even unit test this part and get rid of the explicit file on disk dependency...

KaiSzuttor commented on src/core/reaction_ensemble.cpp:
```diff
@@ -69,26 +78,22 @@ void EnergyCollectiveVariable::load_CV_boundaries(
+    assert(values.size() >= 2);
+    auto it = values.end();
+    wl_system.max_boundaries_energies.push_back(*(--it));
+    wl_system.min_boundaries_energies.push_back(*(--it));
```
```suggestion
    wl_system.max_boundaries_energies.push_back(values.back());
    wl_system.min_boundaries_energies.push_back(values[values.size()-2]));
```

KaiSzuttor commented on src/core/reaction_ensemble.cpp:
```diff
@@ -69,26 +78,22 @@ void EnergyCollectiveVariable::load_CV_boundaries(
 int ReactionAlgorithm::do_reaction(int reaction_steps) {
   for (int i = 0; i < reaction_steps; i++) {
-    int reaction_id = i_random(reactions.size());
+    int reaction_id = i_random(static_cast<int>(reactions.size()));
```
can't you move up the RNG as an arg?

jngrad replied:
```c++
int ReactionAlgorithm::do_reaction(int reaction_steps) {
  printf("======\n");
  for (int i = 0; i < reaction_steps; i++) {
    int reaction_id = i_random(static_cast<int>(reactions.size()));
    printf("%d %zu\n", reaction_id, reactions.size());
    generic_oneway_reaction(reaction_id);
  }
  return 0;
}
```
output:
```
======
0 2
0 2
0 2
0 2
1 2
1 2
1 2
0 2
1 2
0 2
======
0 2
1 2
0 2
0 2
1 2
1 2
0 2
0 2
0 2
1 2
```

KaiSzuttor replied:
What do you want to show here?

jngrad replied:
ok I misunderstood your question. Do you want to pass a RNG lambda function as argument to `do_reaction()`?

KaiSzuttor replied:
ah sorry, maybe it was not clear... yes, i wanted the RNG to be an input from outside of the function such that we could inject the randomness

jngrad replied:
Maybe this can be done in the context of #4154? Right now injecting the randomness wouldn't help with unit testing, because RE classes still rely on the energy functions, that cannot be executed in a unit test.

KaiSzuttor replied:
alright

KaiSzuttor commented on src/core/reaction_ensemble.cpp:
```diff
@@ -972,18 +944,13 @@ double WangLandauReactionEnsemble::calculate_delta_degree_of_association(
+  long needed_bins = 1;
+  // add 1 for degrees of association-related part of histogram (think of only
+  // one acid particle)
+  for (const auto &cv : collective_variables) {
```
this is  a ```std::accumulate```

jngrad replied:
done

KaiSzuttor commented on src/core/reaction_ensemble.cpp:
```diff
@@ -1332,19 +1295,20 @@ void WangLandauReactionEnsemble::write_out_preliminary_energy_run_results(
+      auto const value = static_cast<double>(unraveled_index[i]) *
+                             collective_variables[i]->delta_CV +
+                         collective_variables[i]->CV_minimum;
+      fprintf(pFile, "%f ", value);
```
this function should definitely use a ```ostream```

jngrad replied:
done

KaiSzuttor commented on src/core/unit_tests/reaction_ensemble_classes_test.cpp:
```diff
@@ -0,0 +1,502 @@
+  // exception if not enough particles
+  BOOST_CHECK_THROW(r_algo.measure_excess_chemical_potential(0),
+                    std::runtime_error);
+}
```
this is very verbose for only checking if an exception is thrown

jngrad replied:
test removed

KaiSzuttor commented on testsuite/python/constant_pH_stats.py:
```diff
@@ -0,0 +1,119 @@
+        # relative error
+        self.assertLess(
+            rel_error_alpha,
+            0.015,
```
do we have any idea how the difference depends on given parameters?

jngrad replied:
I'm not sure, and the comment about the pH-pKa difference having to be small in the test setup doesn't seem to always be true, e.g. in the faster test `constant_pH.py ` the pH-pKa difference is larger yet convergence is reached faster. The RNG sequence seems to be more important.

KaiSzuttor reviewed (commented):
Overall looks good, it's just not the prettiest neighbourhood of espresso :P

jngrad committed:
core: Simplify file I/O operations

jngrad committed:
core: Refactor Wang-Landau Collective Variable

jngrad committed:
core: Use std::accumulate

jngrad committed:
tests: Remove WidomInsertion unit test

kodiakhq[bot] committed:
Merge branch 'python' into re_unit_tests
